kernel/comps/framebuffer/src/framebuffer.rs

```diff
@@ -0,0 +1,157 @@
+            );
+            return;
+        }
+    };
+
+    let framebuffer = {
+        let fb_base = framebuffer_arg.address;
+        let fb_size = framebuffer_arg.width
+            * framebuffer_arg.height
+            * (framebuffer_arg.bpp / u8::BITS as usize);
+        let io_mem = IoMem::acquire(fb_base..fb_base + fb_size).unwrap();
+        FrameBuffer {
+            io_mem,
+            width: framebuffer_arg.width,
+            height: framebuffer_arg.height,
+            pixel_format,
+        }
+    };
+
+    framebuffer.clear();
+    FRAMEBUFFER.call_once(|| Arc::new(framebuffer));
+}
+
+impl FrameBuffer {
+    /// Returns the size of the framebuffer in bytes.
+    pub fn size(&self) -> usize {
+        self.io_mem.length()
+    }
+
+    /// Returns the width of the framebuffer in pixels.
+    pub fn width(&self) -> usize {
+        self.width
+    }
+
+    /// Returns the height of the framebuffer in pixels.
+    pub fn height(&self) -> usize {
+        self.height
+    }
+
+    /// Returns the pixel format of the framebuffer.
+    pub fn pixel_format(&self) -> PixelFormat {
+        self.pixel_format
+    }
+
+    /// Returns the rendered pixel.
+    pub fn render_pixel(&self, pixel: Pixel) -> RenderedPixel {
+        pixel.render(self.pixel_format)
+    }
+
+    /// Calculates the offset of a pixel at the specified position.
+    pub fn calc_offset(&self, x: usize, y: usize) -> PixelOffset {
+        PixelOffset {
+            fb: self,
+            offset: (y * self.width + x) * self.pixel_format.nbytes(),
+        }
+    }
+
+    fn check_offset(&self, offset: usize) -> Result<()> {
+        if offset > self.io_mem.length() {
+            log::warn!("Framebuffer offset out of bounds: {}", offset);
+            return Err(Error::AccessDenied);
+        }
+        Ok(())
+    }
+
+    /// Writes a pixel at the specified position.
+    pub fn write_pixel_at(&self, offset: PixelOffset, pixel: RenderedPixel) -> Result<()> {
+        self.check_offset(offset.as_usize() + pixel.nbytes())?;
+        self.io_mem.write_bytes(offset.as_usize(), pixel.as_slice())
+    }
+
+    /// Writes raw bytes at the specified offset.
+    pub fn write_bytes_at(&self, offset: usize, bytes: &[u8]) -> Result<()> {
+        self.check_offset(offset + bytes.len())?;
+        self.io_mem.write_bytes(offset, bytes)
+    }
+
+    /// Clears the framebuffer with default color (black).
+    pub fn clear(&self) {
+        let frame = alloc::vec![0u8; self.size()];
+        self.write_bytes_at(0, &frame).unwrap();
+    }
+}
+
+/// The offset of a pixel in the framebuffer.
+#[derive(Debug, Clone, Copy)]
+pub struct PixelOffset<'a> {
+    fb: &'a FrameBuffer,
+    offset: usize,
+}
+
+impl PixelOffset<'_> {
+    /// Adds the specified delta to the x coordinate.
+    pub fn x_add(&mut self, x_delta: isize) {
+        let delta = x_delta * self.fb.pixel_format.nbytes() as isize;
+        self.offset = self.offset.checked_add_signed(delta).unwrap();
```

Comment:
Any reason why `self.offset += delta` is not good enough?

The debug build always checks for integer overflows.

In the release build, I don't think we need to check for overflows because (1) operating on the pixel offset does not seem to cause overflows very easily, assuming the code is written reasonably, (2) even if it does overflow, it will almost certainly fail the bound check when calling `IoMem::write` or using the offset to index a `Vec`, so it will panic and the problem can be noticed.